### PR TITLE
feat: change Pit output format to XML, close #616

### LINKED ISSUE #616: Use another PIT output format to provide more information
#### Characteristics
- **Issue Type**: feature
- **Severity**: feature

#### Description
Currently DSpot relies on the `CSV` report output produced by PIT. However this report is limited in information. It would be better to use the `XML` report, provided by default, or the `JSON` report provided by Descartes. These two formats include for each mutant the package, class, method name, method signature and line in which the mutant was created. This information could be also included in the report generated by DSpot.

### PR DESCRIPTION
**Pull request by andrewbwogi**
The suggested tasks in #616 have been implemented. It remains to be decided what additional information in the [PitXMLResult](https://github.com/STAMP-project/dspot/compare/master...andrewbwogi:pit-output-format?expand=1#diff-829c10ac7340f0f13c4943d5e66d6f75) should be used in the final DSpot report. If the method signature of the mutated classes will be used, it could be presented in a more intuitive form. Currently it would be in Java bytecode format. @danglotb, what additional information do you think should be used, and is the current format of method signatures ok?

### CODE REVIEW

**Comment by danglotb** [comment]
Hi @andrewbwogi 

Thank you very much for the pull request!

It looks great, I'm gonna review the code and keep you updated! :-)

**Comment by danglotb** [comment]
The additional information seems okay for me. The method descriptor should stay as it is, since it a standard format to describe java methods signatures.

File: dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java [hunk 1]
```diff
@@ -36,19 +35,22 @@
...
+    private Map<CtMethod, Set<AbstractPitResult>> testThatKilledMutants;
 
-    private List<PitResult> mutantNotTestedByOriginal;
+    private List<AbstractPitResult> mutantNotTestedByOriginal;
+
+    private AbstractParser parser;
 
     public PitMutantScoreSelector() {
         this.testThatKilledMutants = new HashMap<>();
+        parser = new PitXMLResultParser();
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java]: Here, the parser is hardcoded. Could we use a command line option or a property to allow the user to use the format he wants, XML or CSV?

File: dspot/src/main/java/eu/stamp_project/utils/pit/AbstractParser.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package eu.stamp_project.utils.pit;
+
+import org.apache.commons.io.FileUtils;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+abstract public class AbstractParser {
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/utils/pit/AbstractParser.java]: We could use a generic here:
```java
abstract public class AbstractParser<T extends  AbstractPitResult>
```

File: dspot/src/main/java/eu/stamp_project/utils/pit/AbstractParser.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            return null;
+        }
+        File directoryReportPit = files[0];
+        if (!directoryReportPit.exists()) {
+            return null;
+        }
+        return new File(directoryReportPit.getPath() + PATH_TO_MUTATIONS_RESULT);
+    }
+
+    public List<AbstractPitResult> parseAndDelete(String pathToDirectoryResults) {
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/utils/pit/AbstractParser.java]: And reuse the generic here:

```java
 public List<T> parseAndDelete(String pathToDirectoryResults) {
```

File: dspot/src/main/java/eu/stamp_project/utils/pit/AbstractParser.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        final List<AbstractPitResult> results = parse(fileResults);
+        try {
+            FileUtils.deleteDirectory(new File(pathToDirectoryResults));
+        } catch (IOException e) {
+            // ignored
+        }
+        return results;
+    }
+
+    abstract public List<AbstractPitResult> parse(File fileResults);
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/utils/pit/AbstractParser.java]: Same here.

File: dspot/src/main/java/eu/stamp_project/utils/pit/PitCSVResult.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package eu.stamp_project.utils.pit;
+
+/**
+ * Created by Benjamin DANGLOT
+ * benjamin.danglot@inria.fr
+ * on 1/4/17
+ */
+public class PitCSVResult extends AbstractPitResult{
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/utils/pit/PitCSVResult.java]: And then, you specify which generic you use:
```java
public class PitCSVResult extends AbstractPitResult<PitCSVResult> {
```

File: dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java [hunk 6]
```diff
@@ -36,19 +35,22 @@
...
     }
 
     public PitMutantScoreSelector(String pathToOriginalResultOfPit) {
         this();
-        initOriginalPitResult(PitResultParser.parse(new File(pathToOriginalResultOfPit)));
+        initOriginalPitResult(parser.parse(new File(pathToOriginalResultOfPit)));
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java]: We may have a problem here. In fact, the `pathToOriginalResultOfPit` come from the command line.

I suggest that: we update the command line options [`(m|--path-pit-result)`](https://github.com/STAMP-project/dspot/blob/master/dspot/src/main/java/eu/stamp_project/utils/options/JSAPOptions.java#L198) in order to:
say that we support also the XML format and according to the given format, we use the correct parser.

Today, one may want to you the CSV format for the initial mutation score but want to use the XML as current output.

**Comment by danglotb** [comment]
@oscarlvp could you check if you need something more?

**Comment by danglotb** [comment]
Any news on the requested changes? Did you encounter any trouble to implement them?

**Comment by andrewbwogi** [comment]
I would like to make some manual tests but I get an error when running the DSpot jar: 
```
no main manifest attribute, in ./dspot-maven-1.2.2-SNAPSHOT.jar
```
I compile and package the project as instructed in the README except I skip tests, but that has not been a problem before. I will try to find whats wrong. Besides testing, your requested changes have been implemented.

**Comment by danglotb** [comment]
Hi @andrewbwogi 

looks great! Gonna check them further.

For the manual testing, what are you doing?

> no main manifest attribute, in ./dspot-maven-1.2.2-SNAPSHOT.jar

Do you try to run the maven plugin with a `java -jar` command line? I think it won't work, but maybe you should just try the dspot core jar, _i.e._ `dspot/target/dspot-1.2.2-SNAPSHOT-jar-with-dependencies.jar` rather than the dspot maven jar, _i.e._ `dspot-maven/target/dspot-maven-1.2.2-SNAPSHOT.jar`.

**Comment by danglotb** [comment]
I'm good with the changes, tell me when it is ready to be merged.

**Comment by danglotb** [comment]
Hi @andrewbwogi 

I do not know this issue. On my side, I cannot reproduce the stuck (DSpot ends well its execution).

According to your logs, PIT is stuck. 

Could you try with: the Jacoco selector? _i.e._ `--test-criterion JacocoCoverageSelector`

Could you also try with the mutation engine `gregor` by using the pit mutation score selector (no flag `--test-criterion` and the flag `--gregor`?

Since I'm not able to reproduce the issue, it is hard to find the problem.

**Comment by andrewbwogi** [comment]
With `--test-criterion JacocoCoverageSelector` DSpot terminates on dhell but not with `--gregor`.

In [test-projects](https://github.com/STAMP-project/dspot/tree/master/dspot/src/test/resources/test-projects), DSpot also freezes with the properties file:

```java
#relative path to the project root from dspot project
project=./
#relative path to the source project from the project properties
src=src/main/java/
#relative path to the test source project from the project properties
testSrc=src/test/java
#java version used
javaVersion=8
#filter used to amplify specific test cases
filter=example.*
#path to the output folder
outputDirectory=target/trash/
# Constant amount of additional time to allow a test to run for
# before considering it to be stuck in an infinite loop
timeoutConstInMillis=10000
#Argument string to use when PIT launches child processes. This is most commonly used
# to increase the amount of memory available to the process,
# but may be used to pass any valid JVM argument.
# Use commas to separate multiple arguments, and put them within brackets
jvmArgs=['-Xmx2048m','-Xms1024m']

#Mutators to apply when using Descartes Mode
#descartesMutators=['void','null','true','false','empty','0','(byte)0','(byte)1','(short)1','(short)2','0L','1L','0.0','1.0','0.0f','1.0f','\\40','\\'A\\'','\\"\\"','\\"A\\"']
descartesMutators=['\\'A\\'']
```
With command:
```shell
java -jar dspot-1.2.2-SNAPSHOT-jar-with-dependencies.jar -p ./test-projects2.properties
````
When I run with `--test-criterion JacocoCoverageSelector` in test-projects I get a java.io.FileNotFoundException. Command:
```shell
java -jar dspot-1.2.2-SNAPSHOT-jar-with-dependencies.jar -p ./test-projects2.properties --test-criterion JacocoCoverageSelector --verbose
````
Execution trace: [out2.txt](https://github.com/STAMP-project/dspot/files/2702609/out2.txt)
Error for this last execution:

> Error: Could not find or load main class ['-Xmx2048m'
> Exception in thread "main" java.lang.RuntimeException: java.io.FileNotFoundException: /home/andrew/Skrivbord/stamp/dspot/dspot/src/test/resources/test-projects/target/dspot/TestListener.ser does not exist.
> 	at eu.stamp_project.testrunner.runner.Loader.load(Loader.java:25)
> 	at eu.stamp_project.testrunner.listener.impl.TestListenerImpl.load(TestListenerImpl.java:113)
> 	at eu.stamp_project.testrunner.EntryPoint.runTests(EntryPoint.java:190)
> 	at eu.stamp_project.testrunner.EntryPoint.runTests(EntryPoint.java:180)
> 	at eu.stamp_project.testrunner.EntryPoint.runTests(EntryPoint.java:148)
> 	at eu.stamp_project.utils.execution.TestRunner.runGivenTestMethods(TestRunner.java:64)
> 	at eu.stamp_project.utils.compilation.TestCompiler.compileAndRun(TestCompiler.java:99)
> 	at eu.stamp_project.utils.compilation.TestCompiler.compileRunAndDiscardUncompilableAndFailingTestMethods(TestCompiler.java:55)
> 	at eu.stamp_project.dspot.Amplification.amplification(Amplification.java:85)
> 	at eu.stamp_project.dspot.DSpot._amplify(DSpot.java:275)
> 	at eu.stamp_project.dspot.DSpot._amplifyTestClass(DSpot.java:266)
> 	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
> 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
> 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
> 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
> 	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
> 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> 	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
> 	at eu.stamp_project.dspot.DSpot._amplifyTestClasses(DSpot.java:262)
> 	at eu.stamp_project.dspot.DSpot.amplifyAllTests(DSpot.java:148)
> 	at eu.stamp_project.Main.run(Main.java:59)
> 	at eu.stamp_project.Main.main(Main.java:41)
> Caused by: java.io.FileNotFoundException: /home/andrew/Skrivbord/stamp/dspot/dspot/src/test/resources/test-projects/target/dspot/TestListener.ser does not exist.
> 	... 22 more


I will try to find at what commit DSpot starts to freeze.

**Comment by danglotb** [comment]
Hi @andrewbwogi 

Where did you find the properties file?

There are at least two errors in it:

```java
jvmArgs=['-Xmx2048m','-Xms1024m']
```

Since 38fe7141385a70c82fd8673fac37c6406f78ddbc, the jvmArgs must not wrapped by brackets, nor in single quote. They must only be separated by commas. You should have:

```java
jvmArgs=-Xmx2048m,-Xms1024m
```

```java
descartesMutators=['\\'A\\'']
```

Same here, you should have:

```java
descartesMutators='A'
```

Here, the single quotes express the fact that the mutator is working on char.

Could you please: 
1. fix your properties and retry?
2. tell me where this properties were? This is a bug in the documentation and we must fix it.

Thank you very much @andrewbwogi ! :-)

**Comment by andrewbwogi** [comment]
The properties file was a copy of a test-projects properties file from an old DSpot commit, I missed the updated syntax. However, even in the current DSpot commit, the use of descartesMutators is specified in the old way with brackets in that file. 

With the updated properties file, DSpot terminates with `--test-criterion JacocoCoverageSelector` but not with `--test-criterion PitMutantScoreSelector`.

**Comment by danglotb** [comment]
Okay this is a great! We merged your pull request before the new year ;)

**Pull request merged**